### DISCUSSION TITLE
Reactors and Projectors not Auto Discovered

### DISCUSSION THREAD

**Discussion by anabeto93** [spatie/laravel-event-sourcing#345] [Q&A] [answered]
This also affects a previous issue that was previously closed. #301 .

Please find a link to a sample app I created with a differently namespaced domain.

https://github.com/anabeto93/spatie-es-auto-discovery

When projectors and reactors are differently namespaced, ie does not follow the default `App\\` pattern as can be seen in `composer.json`, the projectors and reactors are not auto-discovered.

```json
"autoload": {
        "psr-4": {
            "App\\": "app/",
            "Domain\\": "app/Domain/",
            "Database\\Factories\\": "database/factories/",
            "Database\\Seeders\\": "database/seeders/"
        }
    }
```
![different_namespace](https://user-images.githubusercontent.com/7607657/163897974-518d71d2-6d8a-48c2-9b84-1da0f86f9c83.png)

Supplying a path such as `app()->path('Domain')` to `auto_discover_projectors_and_reactors` in the `event-sourcing.php` config file does not lead to auto-discovery of reactors and projectors. The only way to get them to work will be to supply an empty array to `auto_discover_projectors_and_reactors` and then manually register the projectors and reactors as defined in `projectors` and `reactors` properties of the config file.

Doing this still presents a problem where running `php artisan event-sourcing:list` will output that `No Projectors registered` and `No reactors registered` even though these have been manually registered.

![not_discovered](https://user-images.githubusercontent.com/7607657/163898246-1600d6d8-43b2-4f39-b0fc-b39f04ca3477.png)

If the `auto_discover_projectors_and_reactors` property is not left as an empty array and any projector or reactor is manually registered, it'll break the entire application.

![auto-discovery-bug](https://user-images.githubusercontent.com/7607657/163898538-ac4a3362-2abe-4996-bc20-2830f37778fa.png)

If however, the base path to the differently namespaced projectors and reactors are supplied to `auto_discover_projectors_and_reactors` without manually registering any, there will be no errors, however, they will still not be considered registered when running the `php artisan event-sourcing:list` command.

```php
'auto_discover_projectors_and_reactors' => [
        app()->path('Domain/App'),
    ],
```

**Comment by aidan-casey** [comment, marked as answer]
@anabeto93 - In the closed issue that you reference, @brendt requested that you submit a PR with a failing test so that this issue may be addressed. I recommend doing so as opposed to continuing to create issues.

Note that the `auto_discover_base_path` value defines the path that will be stripped from the class name to determine the FQDN. In this scenario, your configuration should be as follows:

```php
<?php

return [

    /*
     * These directories will be scanned for projectors and reactors. They
     * will be registered to Projectionist automatically.
     */
    'auto_discover_projectors_and_reactors' => [
        base_path('app/Domain')
    ],

    /*
     * This directory will be used as the base path when scanning
     * for projectors and reactors.
     */
    'auto_discover_base_path' => base_path('app'),

];
```

**Reply by anabeto93** [reply]
Thank you.
